Improve Append plan when General children exist

prajnamort opened the pull request:
When we append General and Partitioned children together, consider
General as Strewn (rather than SingleQE) to postpone gather motion.
Add a special projection path to General, so that only a single QE
will actually produce rows.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

pivotal-issuemaster commented:
@prajnamort Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=9692)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=9692) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@prajnamort Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=9692)!

kainwen commented:
> Makes sense. Could we do the same for CdbLocusType_Replicated and CdbLocusType_SegmentGeneral?
> 
> In the test with generate_series() that changed its plan, I'm actually not sure if the new plan is better than the old one. It's got an extra Result node, and the output of the generate_series() has to be moved by the Motion, which the old plan would avoid. But I can certainly see that this would be beneficial in more complicated scenarios. Ideally, I guess we would generate both paths, and let the planner choose the cheapest. I don't think we have enough information in this function to make that decision, though; we don't know here where the final result should end up in. So I think this is OK as it is. But let's add another test query that would be more plausibly to benefit from this, maybe something like:
> 
> ```
> postgres=# explain select sum (a) from (select a from dml_union_r where a > 95
> union all
> select g from generate_series(1,2) g) as t;
>                                               QUERY PLAN                                               
> -------------------------------------------------------------------------------------------------------
>  Finalize Aggregate  (cost=26.90..26.91 rows=1 width=8)
>    ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=26.87..26.90 rows=1 width=8)
>          ->  Partial Aggregate  (cost=26.87..26.88 rows=1 width=8)
>                ->  Append  (cost=0.00..14.30 rows=336 width=4)
>                      ->  Seq Scan on dml_union_r  (cost=0.00..4.25 rows=2 width=4)
>                            Filter: (a > 95)
>                      ->  Result  (cost=0.00..20.00 rows=334 width=4)
>                            One-Time Filter: (gp_execution_segment() = 0)
>                            ->  Function Scan on generate_series g  (cost=0.00..10.00 rows=334 width=4)
>  Optimizer: Postgres query optimizer
> (10 rows)
> ```

@prajnamort Thanks for opening this PR.  LGTM overall. 

Hi @hlinnaka 
For replicated table(segment general), it may not be as easy as general because replicated table's
numsegments might be much smaller than the partitioned locus' subpath. Then the result plannode skill cannot be used. And union all a replicated table and partitioned seems a bad SQL programming style.

But as the test cases in this PR shows, general locus might be from dummy path (This is more possible in practical env).

For replicated locus(this is the locus for broadcast motion path), I search the code repo, cannot come up with a case, that the subpath of append path is a path with replicated locus. So optimization for it may not impact practical SQL.

The above is my thoughts. Correct me if I am wrong.

kainwen commented on src/backend/optimizer/util/pathnode.c:
```diff
@@ -1919,14 +1927,39 @@ set_append_path_locus(PlannerInfo *root, Path *pathnode, RelOptInfo *rel,
+				 * will actually produce rows.
+				 */
+				RestrictInfo *restrict_info =
+					make_restrictinfo((Expr *) makeSegmentFilterExpr(0),
```
@prajnamort 
Can the `0` here set a random int?

Something like `random() % (subpath->locus.numsegments)` ?

prajnamort replied:
> @prajnamort
> Can the `0` here set a random int?
> 
> Something like `gp_session_id % (subpath->locus.numsegments)` ?

OK, Done.

kainwen commented on src/test/regress/expected/qp_union_intersect_optimizer.out:
```diff
@@ -1764,6 +1764,80 @@ select g from generate_series(1,2) g;
+                     ->  Seq Scan on dml_union_r
+                           Filter: (a > 95)
+                     ->  Result
+                           One-Time Filter: (gp_execution_segment() = 2)
```
ORCA use `2` here. Could you please run the case many times to see if this will be random?

prajnamort replied:
Yes, I've tested, ORCA randomly pick segment ID.

kainwen commented on src/backend/optimizer/util/pathnode.c:
```diff
@@ -1919,14 +1928,46 @@ set_append_path_locus(PlannerInfo *root, Path *pathnode, RelOptInfo *rel,
+				else
+					numsegments = subpath->locus.numsegments;
+
+				int     segid = gp_session_id % numsegments;
```
We'd better put this var declaration at the beginning of the scope,
or
we just do not use it,  feed `gp_session_id % numsegments` to `makeSegmentFilterExpr` directly.

prajnamort commented:
> Makes sense. Could we do the same for CdbLocusType_Replicated and CdbLocusType_SegmentGeneral?
> 
> In the test with generate_series() that changed its plan, I'm actually not sure if the new plan is better than the old one. It's got an extra Result node, and the output of the generate_series() has to be moved by the Motion, which the old plan would avoid. But I can certainly see that this would be beneficial in more complicated scenarios. Ideally, I guess we would generate both paths, and let the planner choose the cheapest. I don't think we have enough information in this function to make that decision, though; we don't know here where the final result should end up in. So I think this is OK as it is. But let's add another test query that would be more plausibly to benefit from this, maybe something like:
> 
> ```
> postgres=# explain select sum (a) from (select a from dml_union_r where a > 95
> union all
> select g from generate_series(1,2) g) as t;
>                                               QUERY PLAN                                               
> -------------------------------------------------------------------------------------------------------
>  Finalize Aggregate  (cost=26.90..26.91 rows=1 width=8)
>    ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=26.87..26.90 rows=1 width=8)
>          ->  Partial Aggregate  (cost=26.87..26.88 rows=1 width=8)
>                ->  Append  (cost=0.00..14.30 rows=336 width=4)
>                      ->  Seq Scan on dml_union_r  (cost=0.00..4.25 rows=2 width=4)
>                            Filter: (a > 95)
>                      ->  Result  (cost=0.00..20.00 rows=334 width=4)
>                            One-Time Filter: (gp_execution_segment() = 0)
>                            ->  Function Scan on generate_series g  (cost=0.00..10.00 rows=334 width=4)
>  Optimizer: Postgres query optimizer
> (10 rows)
> ```

Hello @hlinnaka 
Thanks for reviewing my PR. I've added the Aggr test you suggested.

For CdbLocusType_Replicated and CdbLocusType_SegmentGeneral:
- I put CdbLocusType_SegmentGeneral under the same logic, it should work just like General here. A dedicated test case are also added.
- But I cannot come up with any test case that append CdbLocusType_Replicated to strewn's.

kainwen commented:
The pipeline failure is not related to this pr.  It passes planner while fails orca. The failure is a known flaky issue and under fixing now.

```
test wrkloadadmin             ... ok (0.62 sec)  (diff:0.12 sec)
parallel group (18 tests):  trig gp_toolkit_ao_funcs expand_table_regression gp_numeric_agg role pg_stat_last_shoperation timeseries runtime_stats pg_stat_last_operation auth_constraint portals_updatable plpgsql_cache expand_table partindex_test expand_table_ao expand_table_aoco partition_pruning
Session terminated, killing shell...make[2]: *** [installcheck-good] Terminated
make[1]: *** [installcheck-regress-recurse] Terminated
make: *** [installcheck-world-src/test-recurse] Terminated
 ...killed.
```
